internal/tasks/manager.go

```diff
@@ -0,0 +1,87 @@
+package tasks
+
+import (
+	"context"
+	"time"
+
+	"github.com/flightctl/flightctl/internal/store"
+	"github.com/flightctl/flightctl/internal/tasks/repotester"
+	"github.com/flightctl/flightctl/pkg/thread"
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+)
+
+type ResourceReference struct {
+	OrgID uuid.UUID
+	Kind  string
+	Name  string
+}
+
+type TaskManager struct {
+	log        logrus.FieldLogger
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	channels   map[string]chan ResourceReference
+	store      store.Store
+}
+
+const (
+	FleetTemplateRollout = "fleet-template-rollout"
+
+	ChannelSize = 20
+)
+
+func Init(log logrus.FieldLogger, store store.Store) TaskManager {
+	ctx := context.Background()
+	ctxWithCancel, cancelFunc := context.WithCancel(ctx)
+
+	channels := make(map[string](chan ResourceReference))
+	channels[FleetTemplateRollout] = make(chan ResourceReference, ChannelSize)
+
+	return TaskManager{
+		log:        log,
+		ctx:        ctxWithCancel,
+		cancelFunc: cancelFunc,
+		channels:   channels,
+		store:      store,
+	}
+}
+
+func (t TaskManager) Start() {
+	repoTester := repotester.NewRepoTester(t.log, t.store)
+	repoTesterThread := thread.New(
+		t.log.WithField("pkg", "repository-tester"), "Repository tester", threadIntervalMinute(2), repoTester.TestRepo)
+	repoTesterThread.Start()
+	defer repoTesterThread.Stop()
+
+	go FleetRollouts(t)
+
+	resourceSync := NewResourceSync(t)
+	resourceSyncThread := thread.New(
+		t.log.WithField("pkg", "resourcesync"), "ResourceSync", threadIntervalMinute(2), resourceSync.Poll)
+	resourceSyncThread.Start()
+	defer resourceSyncThread.Stop()
+}
+
+func (t TaskManager) Stop() {
+	for c := range t.channels {
+		close(t.channels[c])
+	}
+	t.cancelFunc()
```

Comment:
I think you want this order.
- tear down all go routines capable of writing to the channel with cancelFunc()
- close only when you are sure no one can write.

This would panic if stop was called twice so you should doc that or use something like sync.Once https://pkg.go.dev/sync#Once

Reply:
Done